### PR TITLE
engine: adapt engine metrics for multi-rocks

### PR DESCRIPTION
**Pull request by tabokie**
### What is changed and how it works?
Issue Number: Ref #12842 

What's Changed:

- Share one statistics for different tablets
- Iterate all tablets when flushing metrics
- Deprecated enable-statistics for kvdb and raftdb.

```commit-message
None
```

### Related changes

https://github.com/pingcap/docs/pull/11626

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by BusyJay — COMMENTED** [review]
I suggest to always enable statistics to reduce complexity.

File: src/server/engine_factory.rs [hunk 1]
```diff
@@ -38,6 +38,11 @@ pub struct KvEngineFactoryBuilder {
 
 impl KvEngineFactoryBuilder {
     pub fn new(env: Arc<Env>, config: &TikvConfig, cache: Cache) -> Self {
+        let statistics = if config.rocksdb.enable_statistics {
```
[the remark this answers is not in the document]
Comment (tabokie) [reply src/server/engine_factory.rs:41]: It doesn't really simplify much, most places we still need `Option` for engine traits. I'm generally okay but the comment says statistics has quite some overhead:
https://github.com/tikv/tikv/blob/98e8bfb35e6cf87e51916ba229fc269b5c984187/etc/config-template.toml#L563

File: components/server/src/signal_handler.rs [hunk 2]
```diff
@@ -1,18 +1,29 @@
 // Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
 
+use std::sync::Arc;
+
+use engine_rocks::RocksStatistics;
+use engine_traits::{Engines, KvEngine, RaftEngine};
```
[the remark this answers is not in the document]
Comment (tabokie) [reply components/server/src/signal_handler.rs:6]: It's used at L60.

File: src/server/engine_factory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/server/engine_factory.rs:41]: Yes, but the overhead has to be paid. Otherwise we can't provide any meaningful support.

**Comment by BusyJay** [comment]
The issue template should be updated to reflect the configuration change and related doc PR.

**Comment by tabokie** [comment]
Looking good:

image

File: src/server/debug.rs [hunk 3]
```diff
@@ -159,15 +162,41 @@ impl<ER: RaftEngine> Debugger<ER> {
...
         Debugger {
             engines,
+            kv_statistics: None,
+            raft_statistics: None,
             reset_to_version_manager,
             cfg_controller,
         }
     }
 
+    pub fn kv_statistics(&mut self, s: Option<Arc<RocksStatistics>>) {
```
Comment (BusyJay) [inline src/server/debug.rs:172]: Setter should have `set_` prefix unless it's a builder.

File: components/engine_rocks/src/rocks_metrics.rs [hunk 4]
```diff
@@ -910,206 +911,274 @@ pub fn flush_engine_histogram_metrics(t: HistType, value: HistogramData, name: &
...
+}
 
-    // Since block cache is shared, getting cache size from any CF is fine. Here we
-    // get from default CF.
-    let handle = crate::util::get_cf_handle(engine, CF_DEFAULT).unwrap();
-    let block_cache_usage = engine.get_block_cache_usage_cf(handle);
-    STORE_ENGINE_BLOCK_CACHE_USAGE_GAUGE_VEC
-        .with_label_values(&[name, "all"])
-        .set(block_cache_usage as i64);
+pub fn flush_engine_statistics(statistics: &RocksStatistics, name: &str) {
```
Comment (BusyJay) [inline components/engine_rocks/src/rocks_metrics.rs:1162]: I think you can pass a flag to indicate whether include titan.

File: components/engine_rocks/src/rocks_metrics.rs [hunk 5]
```diff
@@ -910,206 +911,274 @@ pub fn flush_engine_histogram_metrics(t: HistType, value: HistogramData, name: &
...
+            cf_stats.obsolete_blob_file_size += db
+                .get_property_int_cf(handle, ROCKSDB_TITANDB_OBSOLETE_BLOB_FILE_SIZE)
+                .unwrap_or(0);
+            cf_stats.blob_file_discardable_ratio_le0 += db
+                .get_property_int_cf(handle, ROCKSDB_TITANDB_DISCARDABLE_RATIO_LE0_FILE)
+                .unwrap_or(0);
+            cf_stats.blob_file_discardable_ratio_le20 += db
+                .get_property_int_cf(handle, ROCKSDB_TITANDB_DISCARDABLE_RATIO_LE20_FILE)
+                .unwrap_or(0);
+            cf_stats.blob_file_discardable_ratio_le50 += db
```
Comment (BusyJay) [inline components/engine_rocks/src/rocks_metrics.rs:1018]: I'm worried about whether 0 is better than `None`, as we have done a lot of work to reduce the size of statistics.

File: src/config/mod.rs [hunk 6]
```diff
@@ -1296,6 +1295,9 @@ impl DbConfig {
             )
             .into());
         }
+        if self.enable_statistics == Some(false) {
+            warn!("raftdb: ignoring `enable_statistics`, statistics is always on.")
```
Comment (BusyJay) [inline src/config/mod.rs:1299]: I think it should be `kvdb`.

File: src/config/mod.rs [hunk 7]
```diff
@@ -1038,7 +1037,9 @@ pub struct DbConfig {
     pub create_if_missing: bool,
     pub max_open_files: i32,
     #[online_config(skip)]
-    pub enable_statistics: bool,
+    #[doc(hidden)]
+    #[serde(skip_serializing)]
+    pub enable_statistics: Option<bool>,
```
Comment (BusyJay) [inline src/config/mod.rs:1042]: Why change to `Option`?

**Review by BusyJay — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**